lib/include/openamp/virtio_mmio.h

```diff
@@ -0,0 +1,485 @@
+
+struct virtio_buf {
+	unsigned int index;
+	unsigned int out_num;
+	unsigned int in_num;
+	struct virtqueue_buf *buffers;
+};
+
+struct virtio_device_mem {
+	void *base;
+	size_t size;
+};
+
+struct virtio_mmio_device {
+	struct virtio_device vdev;
+	struct metal_io_region *cfg_io;
+	struct metal_io_region *shm_io;
+	struct metal_device shm_device;
+	struct virtio_device_mem cfg_mem;
+	struct virtio_device_mem shm_mem;
+	unsigned int device_mode;
+	unsigned int irq;
+	void *user_data;
+};
+
+/**
+ * @brief Get device ID.
+ *
+ * @param[in] dev Pointer to device structure.
+ *
+ * @return Device ID value.
+ */
+
+inline uint32_t virtio_get_devid(const struct virtio_device *vdev)
+{
+	if (!vdev)
+		return 0;
+	return vdev->id.device;
+}
+
+/**
+ * @brief Retrieve device status.
+ *
+ * @param[in] dev Pointer to device structure.
+ *
+ * @return status of the device.
+ */
+
+inline uint8_t virtio_get_status(struct virtio_device *vdev)
+{
+	return vdev->func->get_status(vdev);
+}
+
+/**
+ * @brief Set device status.
+ *
+ * @param[in] dev Pointer to device structure.
+ * @param[in] status Value to be set as device status.
+ *
+ * @return N/A.
+ */
+
+inline void virtio_set_status(struct virtio_device *vdev, uint8_t status)
+{
+	vdev->func->set_status(vdev, status);
+}
+
+/**
+ * @brief Retrieve configuration data from the device.
+ *
+ * @param[in] dev Pointer to device structure.
+ * @param[in] offset Offset of the data within the configuration area.
+ * @param[in] dst Address of the buffer that will hold the data.
+ * @param[in] len Length of the data to be retrieved.
+ *
+ * @return N/A.
+ */
+
+inline void virtio_read_config(struct virtio_device *vdev,
+		 uint32_t offset, void *dst, int length)
+{
+	vdev->func->read_config(vdev, offset, dst, length);
+}
+
+/**
+ * @brief Write configuration data to the device.
+ *
+ * @param[in] dev Pointer to device structure.
+ * @param[in] offset Offset of the data within the configuration area.
+ * @param[in] src Address of the buffer that holds the data to write.
+ * @param[in] len Length of the data to be written.
+ *
+ * @return N/A.
+ */
+inline void virtio_write_config(struct virtio_device *vdev,
+		 uint32_t offset, void *src, int length)
+{
+	vdev->func->write_config(vdev, offset, src, length);
+}
+
+/**
+ * @brief Retrieve features supported by both the VIRTIO driver and the VIRTIO device.
+ *
+ * @param[in] dev Pointer to device structure.
+ *
+ * @return Features supported by both the driver and the device as a bitfield.
+ */
+
+inline uint32_t virtio_get_features(struct virtio_device *vdev)
+{
+	return vdev->func->get_features(vdev);
+}
+
+/**
+ * @brief Set features supported by the VIRTIO driver.
+ *
+ * @param[in] dev Pointer to device structure.
+ * @param[in] features Features supported by the driver as a bitfield.
+ *
+ * @return N/A.
+ */
+
+inline void virtio_set_features(struct virtio_device *vdev, uint32_t features)
+{
+	return vdev->func->set_features(vdev, features);
+}
```

Comment:
these functions above related to the virtio_device struct could be moved in virtio.h ( and some other below)